pkg/controller/direct/directbase/directbase_controller.go

```diff
@@ -0,0 +1,393 @@
+		if err := adapter.Create(ctx, u); err != nil {
+			if unwrappedErr, ok := lifecyclehandler.CausedByUnresolvableDeps(err); ok {
+				logger.Info(unwrappedErr.Error(), "resource", k8s.GetNamespacedName(u))
+				return r.handleUnresolvableDeps(ctx, u, unwrappedErr)
+			}
+			return false, r.handleUpdateFailed(ctx, u, fmt.Errorf("error creating: %w", err))
+		}
+	} else {
+		if _, err = adapter.Update(ctx); err != nil {
+			if unwrappedErr, ok := lifecyclehandler.CausedByUnresolvableDeps(err); ok {
+				logger.Info(unwrappedErr.Error(), "resource", k8s.GetNamespacedName(u))
+				return r.handleUnresolvableDeps(ctx, u, unwrappedErr)
+			}
+			return false, r.handleUpdateFailed(ctx, u, fmt.Errorf("error updating: %w", err))
+		}
+	}
+	if isAPIServerUpdateRequired(u) {
+		return false, r.handleUpToDate(ctx, u)
+	}
+	return false, nil
+}
+
+func (r *reconcileContext) handleUpToDate(ctx context.Context, u *unstructured.Unstructured) error {
+	resource, err := toK8sResource(u)
+	if err != nil {
+		return fmt.Errorf("error converting to k8s resource while handling %v event: %w", k8s.UpToDate, err)
+	}
+	return r.Reconciler.HandleUpToDate(ctx, resource)
+}
+
+func (r *reconcileContext) handleUpdateFailed(ctx context.Context, policy *unstructured.Unstructured, origErr error) error {
+	logger := log.FromContext(ctx)
+
+	resource, err := toK8sResource(policy)
+	if err != nil {
+		logger.Error(err, "error converting to k8s resource while handling event",
+			"resource", k8s.GetNamespacedName(policy), "event", k8s.UpdateFailed)
+		return fmt.Errorf("Update call failed: %w", origErr)
+	}
+	return r.Reconciler.HandleUpdateFailed(ctx, resource, origErr)
+}
+
+func (r *reconcileContext) handleDeleted(ctx context.Context, policy *unstructured.Unstructured) error {
+	resource, err := toK8sResource(policy)
+	if err != nil {
+		return fmt.Errorf("error converting to k8s resource while handling %v event: %w", k8s.Deleted, err)
+	}
+	return r.Reconciler.HandleDeleted(ctx, resource)
+}
+
+func (r *reconcileContext) handleDeleteFailed(ctx context.Context, policy *unstructured.Unstructured, origErr error) error {
+	logger := log.FromContext(ctx)
+
+	resource, err := toK8sResource(policy)
+	if err != nil {
+		logger.Error(err, "error converting to k8s resource while handling event",
+			"resource", k8s.GetNamespacedName(policy), "event", k8s.DeleteFailed)
+		return fmt.Errorf(k8s.DeleteFailedMessageTmpl, origErr)
+	}
+	return r.Reconciler.HandleDeleteFailed(ctx, resource, origErr)
+}
+
+func (r *DirectReconciler) supportsImmediateReconciliations() bool {
+	return r.immediateReconcileRequests != nil
+}
+
+func (r *reconcileContext) handleUnresolvableDeps(ctx context.Context, policy *unstructured.Unstructured, origErr error) (requeue bool, err error) {
+	logger := log.FromContext(ctx)
+
+	resource, err := toK8sResource(policy)
+	if err != nil {
+		return false, fmt.Errorf("error converting to k8s resource while handling unresolvable dependencies event: %w", err)
+	}
```

Comment:
So this is because directbase_controller is currently a fairly direct mapping of the terraform controller.  In about 3 PRs we will switch it to stop doing this k8s.Resource dance, and just work with unstructured.Unstructured.  I'll add a comment & TODO!